src/ApacheOrcDotNet.OptimizedReader/BufferCache.cs

```diff
@@ -0,0 +1,23 @@
+﻿using ApacheOrcDotNet.OptimizedReader.Infrastructure;
+using ApacheOrcDotNet.Protocol;
+using System;
+
+namespace ApacheOrcDotNet.OptimizedReader
+{
+    public class BufferCache
```

Comment:
I don't really like this...  It's not a cache of buffers, it's just a note of the previously used StreamRange.  And it feels hacky plugging into `StreamDetail.StreamKind` like this.  This tells me we're operating at the wrong level of abstracting and this cache probably needs to move up a level.

Reply:
See my last comment for a suggested resolution.

Reply:
Since ranges are per stream, moving it a level up would require testing per stream in places like these:

```
await Task.WhenAll(
    GetByteRangeAsync(streams.Present, _presentStreamCompressedBuffer),
    GetByteRangeAsync(streams.Data, _dataStreamCompressedBuffer),
    GetByteRangeAsync(streams.Secondary, _secondaryStreamCompressedBuffer)
);

DecompressByteRange(streams.Present, _presentStreamCompressedBuffer, ref _presentStreamDecompressedBuffer, ref _presentStreamDecompressedBufferLength);
DecompressByteRange(streams.Data, _dataStreamCompressedBuffer, ref _dataStreamDecompressedBuffer, ref _dataStreamDecompressedBufferLength);
DecompressByteRange(streams.Secondary, _secondaryStreamCompressedBuffer, ref _secondaryStreamDecompressedBuffer, ref _secondaryStreamDecompressedBufferLength);
```

What if I drop the `BufferCache` class and move the logic into the `BaseBuffer` class?

Something like `StreamRange[] _lastStreamByteRanges` and `StreamRange[] _lastStreamDecompressionRanges`?

Reply:
I removed the `BufferCache` class and moved the last used ranges tracking into the `BaseColumnBuffer` class.

I'll give some though on different ways this might be cached, but using the `StreamDetail` record would require some refactoring since the `StreamDetail` is cached in the reader and we only process the `Positions` and `Range` in `GetColumnDataStreams` (for each available stream, per row entry index).

P.S.: I ran all the file read tests again and the extra `10ms` we observed was due to some test code that I had before the main loop:

```
reader.LoadDataAsync(0, 65, timeColumnBuffer).Wait();
reader.LoadDataAsync(0, 66, timeColumnBuffer).Wait();
reader.LoadDataAsync(0, 67, timeColumnBuffer).Wait();
```

Removing those got us back into an average `90ms` execution time with the test dataset.